### PR TITLE
Add action auto-scroll

### LINKED ISSUE #25186: [Actions] Auto scroll feature
### Feature Description

It would be nice if on the action page it would auto scroll down to the bottom of the logs (and maybe auto open the steps?)

### Screenshots

_No response_

### LINKED ISSUE #28535: Add a "Follow" or "Auto-scroll" button to the actions steps
### Feature Description

Many times a given step in a job will have a lot of output.  `molecule test` is a good example ...  Currently, as new output appears it's added to the end of the step output, but you can't see it unless you manually scroll the window.

It would be really nice to have a button on the step line when it's opened up to toggle auto-scrolling on and off.

For example **Drone** CI uses an eye symbol for this functionality
image
That can be toggled on and off
image
Oddly, the 2nd one, strike-through-eye is the one for auto-scrolling, I would have thought that meant "stop scrolling" but OK ...

Personally I think the toggle button could also change the background color of the step line to indicate active or not.  That's more quickly caught by the eyes than an icon - but also use a toggle icon ...

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by bytedream**
Adds an auto-scroll/follow feature to running actions (fix #25186, fix #28535).

When new log lines are appended and the bottom of the logs container (`.action-view-right`) is visible at this time, the page automatically scrolls down to the bottom of the logs.

### CODE REVIEW

File: web_src/js/components/RepoActionView.vue [hunk 1]
```diff
@@ -188,11 +188,22 @@ const sfc = {
...
       for (const line of logLines) {
         // TODO: group support: ##[group]GroupTitle , ##[endgroup]
         const el = this.getLogsContainer(stepIndex);
         el.append(this.createLogLine(line, startTime, stepIndex));
       }
+
+      // scrolls to the bottom if job is running and the bottom of the logs container is visible
+      if (!this.run.done && logLines.length > 0 && clientHeight >= logsContainerHeight) {
+        const newLogsContainerHeight = this.$refs.stepsContainer.getBoundingClientRect().bottom + window.scrollY;
+        window.scrollTo({top: clientHeight + (newLogsContainerHeight - logsContainerHeight), behavior: 'smooth'});
```
Comment (silverwind) [inline web_src/js/components/RepoActionView.vue]: Does it have to be `smooth`? Personally I hate all kinds of smooth scrolling. We could set `auto` here which would then enable themes to change the behaviour via CSS [`scroll-behaviour`](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-behavior) if they so desire.

File: web_src/js/components/RepoActionView.vue [hunk 2]
```diff
@@ -188,11 +188,22 @@ const sfc = {
...
+      const logsContainerHeight = this.$refs.stepsContainer.getBoundingClientRect().bottom + window.scrollY;
+
       for (const line of logLines) {
         // TODO: group support: ##[group]GroupTitle , ##[endgroup]
         const el = this.getLogsContainer(stepIndex);
         el.append(this.createLogLine(line, startTime, stepIndex));
       }
+
+      // scrolls to the bottom if job is running and the bottom of the logs container is visible
+      if (!this.run.done && logLines.length > 0 && clientHeight >= logsContainerHeight) {
```
Comment (silverwind) [inline web_src/js/components/RepoActionView.vue]:
```suggestion
      if (!this.run.done && logLines.length && clientHeight >= logsContainerHeight) {
```

**Comment by silverwind** [comment]
This kind of scrolling while nice, does present a UX issue when the user scrolls up: It would scroll it down again when the next line is received which is disruptive. It would be better to:

- Scroll down as long as the user has not manually scrolled
- If the user has manually scrolled and is not at the end, don't scroll
- If the user has manually scrolled and is at the end, scroll

That way, a user can to view and for example copy lines in a running step, while still having autoscroll when they scroll to bottom.

File: web_src/js/components/RepoActionView.vue (hunk 1 above)
Comment (bytedream) [reply web_src/js/components/RepoActionView.vue]: With smooth scrolling it's clearer to see when new lines were added, especially when the newly added lines are looking similar. I often have action logs on my second screen to see if the changes I've made are passing while I continue working on my main screen, and with smooth scrolling, I notice new logs faster.

<details>
<summary>Videos of both scroll behaviors</summary>

https://github.com/go-gitea/gitea/assets/63594396/ed9acecd-7bc3-44a9-a141-c4f614837642

https://github.com/go-gitea/gitea/assets/63594396/7e61b5b8-e82b-48d1-a1f3-142d4e9a3460
</details>

Comment (silverwind) [reply web_src/js/components/RepoActionView.vue]: I guess it's an argument for it, yes.

**Comment by bytedream** [comment]
I try to comply to this behavior by only scrolling if the bottom of the container which contains all job steps is visible when new lines are added. So if the user scrolls up and the container bottom isn't visible, it won't continue scrolling.
This was the solution with the fewest code changes I came up with, but I'm also fine with the behavior you suggested.

**Comment by silverwind** [comment]
Sounds good. I guess this case is a bit tricker than the ones I had because it's not a simple self-contained element that scrolls but the whole page. When in doubt, check how GitHub does it.

**Comment by bytedream** [comment]
GitHub has the same behavior as my PR as long as there are only a few steps (+ the scroll down when visiting the an action page).
When an action has many steps which may exceed the screen height, GitHub keeps the logs around the middle of the screen, if the following to-be-executed job steps are still exceeding the remaining screen height. When manually scrolling further down so that the latest log line is in the upper half of the screen (or completely out of view), GitHub scrolls up again to keep the latest logs in the middle of the screen.
For example, [Rust](https://github.com/rust-lang/rust/actions) has actions with many steps.

From how GitHub does it, how you suggested it and how I would like it, I purpose the following:
- Scroll down as long as the user has not manually scrolled
- When the user has manually scrolled, keep following the logs when the last log line is in the bottom half of the screen, else do not follow

**Comment by wxiaoguang** [comment]
It needs some improvements, I will make some changes in this PR

**Comment by wxiaoguang** [comment]
Made some changes in 5d0ecccbdd5ea716b9426e8ee4bce9332a36de7f, now we first collect all the auto-scroll-able log containers, and then scroll the last one.

It should also be easier to add a "toggle" to switch the auto-scroll in `shouldAutoScroll`, but I think we do not need to do so at the moment, the default behavior seems good enough, then we do not need to make the UI more complex.

**Review by wxiaoguang — APPROVED** [review]

**Comment by wxiaoguang** [comment]
I have made a mock endpoint `http://localhost:3000/devtest/repo-action-view`, it could test the Vue component behaviors easily without really setting up an action.

**Review by lunny — APPROVED** [review]

**Pull request merged**